dotrix_core/src/renderer/texture.rs

```diff
@@ -183,4 +292,90 @@ impl Texture {
     pub fn is_filterable(&self) -> bool {
         self.format.describe().guaranteed_format_features.filterable
     }
+
+    /// Get the texture bytes per pixels
+    pub fn pixel_bytes(&self) -> u8 {
+        self.format.describe().block_size
+    }
+
+    /// Get the number of channels
+    pub fn num_channels(&self) -> u8 {
+        self.format.describe().components
+    }
+
+    /// Get the texture sample type (float/uint etc)
+    pub fn sample_type(&self) -> wgpu::TextureSampleType {
+        self.format.describe().sample_type
+    }
+
+    /// Fetch data from the gpu
+    ///
+    /// This is useful textures that are altered on the gpu
+    ///
+    /// This operation is slow and should mostly be
+    /// used for debugging
+    pub fn fetch_from_gpu(
+        &self,
+        dimensions: [u32; 3],
+        ctx: &mut Context,
+    ) -> impl std::future::Future<Output = Result<Vec<u8>, wgpu::BufferAsyncError>> {
```

Comment:
This is a convenience method to get a texture into a `Vec<u8>`.  It must be async because we must wait for the commands to be enqueued and run on the gpu.

This is an example function I used for debug printing a texture


```rust
fn print_texture(texture: &TextureBuffer, dimensions: [u32; 3], renderer: &mut Renderer) {
    let future = renderer.fetch_texture(texture, dimensions);
    let channels = texture.num_channels() as u32;
    let bytes_per_pixel = texture.pixel_bytes() as u32;
    std::thread::spawn(move || {
        futures::executor::block_on(async {
            if let Ok(data) = future.await {
                // Since contents are in bytes, this converts these bytes back to f32
                // strips the padding and puts it into 3d format
                let bytes_per_row = dimensions[0] * bytes_per_pixel;
                let result: Vec<Vec<Vec<Vec<f32>>>> = data
                    .chunks_exact((bytes_per_row * dimensions[1]) as usize)
                    .map(|img| {
                        let rows: Vec<Vec<Vec<f32>>> = img
                            .chunks_exact(bytes_per_row as usize)
                            .map(|row| {
                                let row_f32: Vec<f32> = row
                                    .chunks_exact(4)
                                    .map(|b| f32::from_ne_bytes(b.try_into().unwrap()))
                                    .collect();
                                let pixels: Vec<Vec<f32>> = row_f32
                                    .chunks(channels as usize)
                                    .map(|pixels| pixels.to_vec())
                                    .collect();
                                pixels[0..(dimensions[0] as usize)].to_vec()
                            })
                            .collect();
                        rows
                    })
                    .collect();

                for (idx, img) in result.iter().enumerate() {
                    println!("Z={}", idx);
                    for row in img.iter() {
                        println!("{:.2?}", row);
                    }
                }
            }
        })
    });
}
```

It converts the `u8` bytes to `f32` because the texture I used used full f32 and should be adapted for other usages

Reply:
I currently do not have a good way to ensure that the commands have been enqueued and run and just wait for 1.0 seconds. A way to get the cycle count asyncronously and wait for it might be useful here